### PR TITLE
NameCache refactoring

### PR DESCRIPTION
**Pull request by hsyl20**
Companion MR for https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4951

### CODE REVIEW

**Comment by Kleidukos** [comment]
oooh I like this kind of diff! :)

**Review by Kleidukos — APPROVED** [review]

**Comment by alexbiehl** [comment]
Thank you sir! 🚢

**Review by alexbiehl — APPROVED** [review]

File: hypsrc-test/ref/src/Classes.html [hunk 1]
```diff
@@ -194,7 +194,7 @@
 	><span class="annottext"
 	  >bar :: Int -&gt; Int
 </span
-	  ><a href="#"
+	  ><a href="Classes.html#bar"
```
Comment (alexbiehl) [inline hypsrc-test/ref/src/Classes.html:197]: I like it, but I don't know why your patch could cause it.

Comment (hsyl20) [reply hypsrc-test/ref/src/Classes.html:197]: I have no idea either. It appeared recently after a rebase

**Pull request merged**